src/Processors/Formats/Impl/BSONEachRowRowInputFormat.cpp

```diff
@@ -0,0 +1,965 @@
+            in.ignore(size);
+            switch (subtype)
+            {
+                case BSONBinarySubtype::BINARY_OLD: [[fallthrough]];
+                case BSONBinarySubtype::BINARY:
+                    return makeNullable(std::make_shared<DataTypeString>());
+                case BSONBinarySubtype::UUID_OLD: [[fallthrough]];
+                case BSONBinarySubtype::UUID:
+                    return makeNullable(std::make_shared<DataTypeUUID>());
+                default:
+                    throw Exception(ErrorCodes::UNKNOWN_TYPE, "BSON binary subtype {} is not supported", getBSONBinarySubtypeName(subtype));
+            }
+        }
+        case BSONType::NULL_VALUE:
+        {
+            return nullptr;
+        }
+        default:
+        {
+            if (!allow_to_skip_unsupported_types)
+                throw Exception(ErrorCodes::UNKNOWN_TYPE, "BSON type {} is not supported", getBSONTypeName(type));
+
+            skip = true;
+            skipBSONField(in, type);
+            return nullptr;
+        }
+    }
+}
+
+NamesAndTypesList BSONEachRowSchemaReader::getDataTypesFromBSONDocument(bool allow_to_skip_unsupported_types)
+{
+    size_t document_start = in.count();
+    BSON_SIZE_TYPE document_size;
+    readBinary(document_size, in);
+    NamesAndTypesList names_and_types;
+    while (in.count() - document_start + sizeof(BSON_DOCUMENT_END) != document_size)
+    {
+        auto bson_type = getBSONType(readBSONType(in));
+        String name;
+        readNullTerminated(name, in);
+        bool skip = false;
+        auto type = getDataTypeFromBSONField(bson_type, allow_to_skip_unsupported_types, skip);
+        if (!skip)
+            names_and_types.emplace_back(name, type);
+    }
+
+    assertChar(BSON_DOCUMENT_END, in);
+
+    return names_and_types;
+}
+
+NamesAndTypesList BSONEachRowSchemaReader::readRowAndGetNamesAndDataTypes(bool & eof)
+{
+    if (in.eof())
+    {
+        eof = true;
+        return {};
+    }
+
+    return getDataTypesFromBSONDocument(format_settings.bson.skip_fields_with_unsupported_types_in_schema_inference);
+}
+
+void BSONEachRowSchemaReader::transformTypesIfNeeded(DataTypePtr & type, DataTypePtr & new_type)
+{
+    DataTypes types = {type, new_type};
+    /// For example for integer conversion Int32,
+    auto least_supertype = tryGetLeastSupertype(types);
+    if (least_supertype)
+        type = new_type = least_supertype;
+}
+
+static std::pair<bool, size_t>
+fileSegmentationEngineBSONEachRow(ReadBuffer & in, DB::Memory<> & memory, size_t min_bytes, size_t max_rows)
+{
+    size_t number_of_rows = 0;
+
+    while (!in.eof() && memory.size() < min_bytes && number_of_rows < max_rows)
+    {
+        BSON_SIZE_TYPE document_size;
+        readBinary(document_size, in);
+        size_t old_size = memory.size();
+        memory.resize(old_size + document_size);
+        memcpy(memory.data() + old_size, reinterpret_cast<char *>(&document_size), sizeof(document_size));
```

Comment:
Good remark. Also in BSON format all types are always serialized in little-endian order, so this format won't work at all in big-endian platforms. I will add a remark in documentation. I don't think we should change this place because ayway we will read document size in a wrong order and segmantation won't work